File: lib/services/telegram_manager.py
import os
import logging
import requests
import asyncio
from typing import List, Dict, Optional, Any
from dotenv import load_dotenv

class TelegramManager:
    """
    Kelas untuk mengelola sesi dan interaksi dengan Telegram
    """
    
    def __init__(self):
        # Setup logging
        self.logger = logging.getLogger('TelegramManager')
        self.logger.setLevel(logging.INFO)
        
        # Konfigurasi handler dan formatter
        handler = logging.StreamHandler()
        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
        handler.setFormatter(formatter)
        self.logger.addHandler(handler)
        
        # Konfigurasi base URL dan session
        self.base_url = 'http://localhost:5000'  # URL server backend
        self.session = requests.Session()
        self.phone_number = None
        
        # Load environment variables
        load_dotenv()

        self.selected_session = None
        
        # Pastikan folder sessions ada
        if not os.path.exists('sessions'):
            os.makedirs('sessions')

    async def _make_request(self, method: str, endpoint: str, data: dict = None) -> dict:
        """Membuat HTTP request dengan async"""
        try:
            loop = asyncio.get_event_loop()
            if method.upper() == 'GET':
                response = await loop.run_in_executor(
                    None, 
                    lambda: self.session.get(f'{self.base_url}/{endpoint}')
                )
            else:  # POST
                response = await loop.run_in_executor(
                    None, 
                    lambda: self.session.post(f'{self.base_url}/{endpoint}', json=data)
                )
            
            response.raise_for_status()
            return response.json()
        except Exception as e:
            self.logger.error(f'Error making request to {endpoint}: {str(e)}')
            raise

    async def create_session(self, phone_number: str) -> None:
        """Membuat sesi baru"""
        try:
            data = await self._make_request('POST', 'create_session', {'phone_number': phone_number})
            if data['status'] == 'success':
                self.phone_number = phone_number
            else:
                raise Exception(data['message'])
        except Exception as e:
            self.logger.error(f'Error creating session: {str(e)}')
            raise

    async def verify_code(self, code: str) -> None:
        """Verifikasi kode"""
        if not self.phone_number:
            raise Exception('Silakan kirim kode verifikasi terlebih dahulu')
            
        try:
            data = await self._make_request('POST', 'verify_code', {
                'phone_number': self.phone_number,
                'code': code
            })
            
            if data['status'] != 'success':
                raise Exception(data['message'])
        except Exception as e:
            self.logger.error(f'Error verifying code: {str(e)}')
            raise

    async def get_chat_history(self) -> List[str]:
        """Mendapatkan riwayat chat"""
        if not self.selected_session:
            raise Exception('Silakan pilih sesi terlebih dahulu')
            
        try:
            data = await self._make_request('POST', 'chat_history', {
                'phone_number': self.selected_session
            })
            
            if data['status'] == 'success':
                return data['chat_history']
            else:
                raise Exception(data['message'])
        except Exception as e:
            self.logger.error(f'Error getting chat history: {str(e)}')
            raise

    async def change_username(self, username: str) -> None:
        """Mengubah username"""
        if not self.selected_session:
            raise Exception('Silakan pilih sesi terlebih dahulu')
            
        try:
            data = await self._make_request('POST', 'change_username', {
                'phone_number': self.selected_session,
                'username': username
            })
            
            if data['status'] != 'success':
                raise Exception(data['message'])
        except Exception as e:
            self.logger.error(f'Error changing username: {str(e)}')
            raise

    async def change_name(self, first_name: str, last_name: Optional[str] = None) -> None:
        """Mengubah nama profil"""
        if not self.selected_session:
            raise Exception('Silakan pilih sesi terlebih dahulu')
            
        try:
            data = await self._make_request('POST', 'change_name', {
                'phone_number': self.selected_session,
                'first_name': first_name,
                'last_name': last_name
            })
            
            if data['status'] != 'success':
                raise Exception(data['message'])
        except Exception as e:
            self.logger.error(f'Error changing name: {str(e)}')
            raise

    async def list_channels(self) -> List[Dict[str, Any]]:
        """Mendapatkan daftar channel dan grup"""
        if not self.selected_session:
            raise Exception('Silakan pilih sesi terlebih dahulu')
            
        try:
            data = await self._make_request('POST', 'list_channels', {
                'phone_number': self.selected_session
            })
            
            if data['status'] == 'success':
                return data['channels']
            else:
                raise Exception(data['message'])
        except Exception as e:
            self.logger.error(f'Error listing channels: {str(e)}')
            raise

    async def leave_all_channels(self) -> None:
        """Keluar dari semua channel dan grup"""
        if not self.selected_session:
            raise Exception('Silakan pilih sesi terlebih dahulu')
            
        try:
            data = await self._make_request('POST', 'leave_all_channels', {
                'phone_number': self.selected_session
            })
            
            if data['status'] != 'success':
                raise Exception(data['message'])
        except Exception as e:
            self.logger.error(f'Error leaving channels: {str(e)}')
            raise

    async def delete_session(self, phone_number: str) -> None:
        """Menghapus sesi"""
        try:
            data = await self._make_request('POST', 'delete_session', {
                'phone_number': phone_number
            })
            
            if data['status'] == 'success':
                if self.selected_session == phone_number:
                    self.selected_session = None
                if self.phone_number == phone_number:
                    self.phone_number = None
            else:
                raise Exception(data['message'])
        except Exception as e:
            self.logger.error(f'Error deleting session: {str(e)}')
            raise 

    async def list_sessions(self) -> List[str]:
        """Mendapatkan daftar sesi yang tersedia"""
        try:
            response = await self._make_request('GET', 'list_sessions')
            if response['status'] == 'success':
                return response['sessions']
            else:
                raise Exception(response['message'])
        except Exception as e:
            self.logger.error(f'Error listing sessions: {str(e)}')
            raise 

    async def select_session(self, phone_number: str) -> None:
        """Memilih sesi yang akan digunakan"""
        try:
            # Cek apakah file sesi ada
            session_file = os.path.join('sessions', f"{phone_number}.session")
            if not os.path.exists(session_file):
                raise Exception(f"Sesi untuk nomor {phone_number} tidak ditemukan")
            
            # Set sesi yang dipilih
            self.selected_session = phone_number
            self.logger.info(f"Selected session: {phone_number}")
        except Exception as e:
            self.logger.error(f"Error selecting session: {str(e)}")
            raise 

File: lib/services/telegram_manager.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:dio/dio.dart';
import 'package:logging/logging.dart';

class TelegramManager {
  static final _logger = Logger('TelegramManager');
  static const _baseUrl = 'http://192.168.124.233:5000';
  
  final _storage = const FlutterSecureStorage();
  final _dio = Dio();
  String? _phoneNumber;

  Future<void> initialize() async {
    _logger.info('Initializing TelegramManager');
  }

  Future<List<String>> listSessions() async {
    try {
      _logger.info('Mengambil daftar sesi dari $_baseUrl/list_sessions');
      
      final response = await _dio.get(
        '$_baseUrl/list_sessions',
        options: Options(
          headers: {
            'Accept': 'application/json',
          },
          validateStatus: (status) => true,
        ),
      );

      _logger.info('Response status: ${response.statusCode}');
      _logger.info('Response data: ${response.data}');

      if (response.statusCode != 200) {
        throw Exception(response.data['message'] ?? 'Unknown error');
      }

      if (response.data['status'] == 'success') {
        return List<String>.from(response.data['sessions']);
      } else {
        throw Exception(response.data['message']);
      }
    } catch (e) {
      _logger.severe('Error listing sessions: $e');
      rethrow;
    }
  }

  Future<void> selectSession(String phoneNumber) async {
    _phoneNumber = phoneNumber;
    await _storage.write(
      key: 'selected_session',
      value: phoneNumber,
    );
    _logger.info('Selected session: $phoneNumber');
  }

  Future<String?> getSelectedSession() async {
    return await _storage.read(key: 'selected_session');
  }

  Future<void> createSession(String phoneNumber) async {
    await initialize();
    _phoneNumber = phoneNumber;
    
    try {
      _logger.info('Mengirim request ke $_baseUrl/create_session');
      _logger.info('Phone number: $phoneNumber');
      
      final response = await _dio.post(
        '$_baseUrl/create_session',
        data: {
          'phone_number': phoneNumber,
        },
        options: Options(
          headers: {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
          },
          validateStatus: (status) => true,
        ),
      );

      _logger.info('Response status: ${response.statusCode}');
      _logger.info('Response data: ${response.data}');

      if (response.statusCode != 200) {
        throw Exception(response.data['message'] ?? 'Unknown error');
      }

      if (response.data['status'] == 'success') {
        _logger.info('Kode verifikasi telah dikirim');
      } else {
        throw Exception(response.data['message']);
      }
    } catch (e) {
      _logger.severe('Error creating session: $e');
      rethrow;
    }
  }

  Future<void> verifyCode(String code) async {
    if (_phoneNumber == null) {
      throw Exception('Silakan kirim kode verifikasi terlebih dahulu');
    }

    try {
      _logger.info('Mengirim request ke $_baseUrl/verify_code');
      _logger.info('Phone number: $_phoneNumber');
      _logger.info('Code: $code');

      final response = await _dio.post(
        '$_baseUrl/verify_code',
        data: {
          'phone_number': _phoneNumber,
          'code': code,
        },
        options: Options(
          headers: {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
          },
          validateStatus: (status) => true,
        ),
      );

      _logger.info('Response status: ${response.statusCode}');
      _logger.info('Response data: ${response.data}');

      if (response.statusCode != 200) {
        throw Exception(response.data['message'] ?? 'Unknown error');
      }

      if (response.data['status'] == 'success') {
        _logger.info('Sesi berhasil dibuat');
        await _storage.write(
          key: 'session_$_phoneNumber',
          value: _phoneNumber,
        );
      } else {
        throw Exception(response.data['message']);
      }
    } catch (e) {
      _logger.severe('Error verifying code: $e');
      rethrow;
    }
  }

  Future<List<String>> getChatHistory() async {
    if (_phoneNumber == null) {
      throw Exception('Silakan buat sesi terlebih dahulu');
    }

    try {
      _logger.info('Mengirim request ke $_baseUrl/chat_history');
      _logger.info('Phone number: $_phoneNumber');

      final response = await _dio.post(
        '$_baseUrl/chat_history',
        data: {
          'phone_number': _phoneNumber,
        },
        options: Options(
          headers: {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
          },
          validateStatus: (status) => true,
        ),
      );

      _logger.info('Response status: ${response.statusCode}');
      _logger.info('Response data: ${response.data}');

      if (response.statusCode != 200) {
        throw Exception(response.data['message'] ?? 'Unknown error');
      }

      if (response.data['status'] == 'success') {
        _logger.info('Riwayat chat berhasil diambil');
        return List<String>.from(response.data['chat_history']);
      } else {
        throw Exception(response.data['message']);
      }
    } catch (e) {
      _logger.severe('Error getting chat history: $e');
      rethrow;
    }
  }

  Future<void> changeUsername(String username) async {
    if (_phoneNumber == null) {
      throw Exception('Silakan buat sesi terlebih dahulu');
    }

    try {
      _logger.info('Mengirim request ke $_baseUrl/change_username');
      _logger.info('Phone number: $_phoneNumber');
      _logger.info('New username: $username');

      final response = await _dio.post(
        '$_baseUrl/change_username',
        data: {
          'phone_number': _phoneNumber,
          'username': username,
        },
        options: Options(
          headers: {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
          },
          validateStatus: (status) => true,
        ),
      );

      _logger.info('Response status: ${response.statusCode}');
      _logger.info('Response data: ${response.data}');

      if (response.statusCode != 200) {
        throw Exception(response.data['message'] ?? 'Unknown error');
      }

      if (response.data['status'] == 'success') {
        _logger.info('Username berhasil diubah');
      } else {
        throw Exception(response.data['message']);
      }
    } catch (e) {
      _logger.severe('Error changing username: $e');
      rethrow;
    }
  }

  Future<void> changeName(String firstName, String? lastName) async {
    if (_phoneNumber == null) {
      throw Exception('Silakan buat sesi terlebih dahulu');
    }

    try {
      _logger.info('Mengirim request ke $_baseUrl/change_name');
      _logger.info('Phone number: $_phoneNumber');
      _logger.info('New name: $firstName $lastName');

      final response = await _dio.post(
        '$_baseUrl/change_name',
        data: {
          'phone_number': _phoneNumber,
          'first_name': firstName,
          'last_name': lastName,
        },
        options: Options(
          headers: {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
          },
          validateStatus: (status) => true,
        ),
      );

      _logger.info('Response status: ${response.statusCode}');
      _logger.info('Response data: ${response.data}');

      if (response.statusCode != 200) {
        throw Exception(response.data['message'] ?? 'Unknown error');
      }

      if (response.data['status'] == 'success') {
        _logger.info('Nama berhasil diubah');
      } else {
        throw Exception(response.data['message']);
      }
    } catch (e) {
      _logger.severe('Error changing name: $e');
      rethrow;
    }
  }

  Future<List<Map<String, dynamic>>> listChannels() async {
    if (_phoneNumber == null) {
      throw Exception('Silakan buat sesi terlebih dahulu');
    }

    try {
      _logger.info('Mengirim request ke $_baseUrl/list_channels');
      _logger.info('Phone number: $_phoneNumber');

      final response = await _dio.post(
        '$_baseUrl/list_channels',
        data: {
          'phone_number': _phoneNumber,
        },
        options: Options(
          headers: {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
          },
          validateStatus: (status) => true,
        ),
      );

      _logger.info('Response status: ${response.statusCode}');
      _logger.info('Response data: ${response.data}');

      if (response.statusCode != 200) {
        throw Exception(response.data['message'] ?? 'Unknown error');
      }

      if (response.data['status'] == 'success') {
        _logger.info('Daftar channel berhasil diambil');
        return List<Map<String, dynamic>>.from(response.data['channels']);
      } else {
        throw Exception(response.data['message']);
      }
    } catch (e) {
      _logger.severe('Error listing channels: $e');
      rethrow;
    }
  }

  Future<void> leaveAllChannels() async {
    if (_phoneNumber == null) {
      throw Exception('Silakan buat sesi terlebih dahulu');
    }

    try {
      _logger.info('Mengirim request ke $_baseUrl/leave_all_channels');
      _logger.info('Phone number: $_phoneNumber');

      final response = await _dio.post(
        '$_baseUrl/leave_all_channels',
        data: {
          'phone_number': _phoneNumber,
        },
        options: Options(
          headers: {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
          },
          validateStatus: (status) => true,
        ),
      );

      _logger.info('Response status: ${response.statusCode}');
      _logger.info('Response data: ${response.data}');

      if (response.statusCode != 200) {
        throw Exception(response.data['message'] ?? 'Unknown error');
      }

      if (response.data['status'] == 'success') {
        _logger.info('Berhasil keluar dari semua channel');
      } else {
        throw Exception(response.data['message']);
      }
    } catch (e) {
      _logger.severe('Error leaving channels: $e');
      rethrow;
    }
  }

  Future<void> deleteSession(String phoneNumber) async {
    try {
      _logger.info('Mengirim request ke $_baseUrl/delete_session');
      _logger.info('Phone number: $phoneNumber');

      final response = await _dio.post(
        '$_baseUrl/delete_session',
        data: {
          'phone_number': phoneNumber,
        },
        options: Options(
          headers: {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
          },
          validateStatus: (status) => true,
        ),
      );

      _logger.info('Response status: ${response.statusCode}');
      _logger.info('Response data: ${response.data}');

      if (response.statusCode != 200) {
        throw Exception(response.data['message'] ?? 'Unknown error');
      }

      if (response.data['status'] == 'success') {
        _logger.info('Sesi berhasil dihapus');
        // Hapus sesi dari penyimpanan lokal jika itu sesi yang aktif
        final selectedSession = await getSelectedSession();
        if (selectedSession == phoneNumber) {
          await _storage.delete(key: 'selected_session');
          _phoneNumber = null;
        }
        await _storage.delete(key: 'session_$phoneNumber');
      } else {
        throw Exception(response.data['message']);
      }
    } catch (e) {
      _logger.severe('Error deleting session: $e');
      rethrow;
    }
  }
} 